pdf/pdfselection.cpp

```diff
@@ -0,0 +1,486 @@
+    } else {
+        int n = -boxIndex2;
+        for (int i = pageIndex2; i < pageIndex1; i++) {
+            n += doc->textBoxesAtPage(i).count();
+        }
+        return -n - boxIndex1;
+    }
+}
+
+int PDFSelection::count() const
+{
+    if (d->pageIndexStart < 0 || d->pageIndexStop < 0 ||
+        d->boxIndexStart < 0 || d->boxIndexStop < 0) {
+        return 0;
+    }
+
+    if (!d->canvas || !d->canvas->document())
+        return 0;
+
+    return d->sliceCount(d->pageIndexStart, d->boxIndexStart,
+                         d->pageIndexStop, d->boxIndexStop) + 1;
+}
+
+PDFCanvas* PDFSelection::canvas() const
+{
+    return d->canvas;
+}
+
+void PDFSelection::setCanvas(PDFCanvas *newCanvas)
+{
+    if (newCanvas != d->canvas) {
+        if (d->canvas)
+            d->canvas->disconnect(this);
+        d->canvas = newCanvas;
+
+        connect(d->canvas, &PDFCanvas::pageLayoutChanged, this, &PDFSelection::handle1Changed);
+        connect(d->canvas, &PDFCanvas::pageLayoutChanged, this, &PDFSelection::handle2Changed);
+        connect(d->canvas, &PDFCanvas::pageLayoutChanged, this, &PDFSelection::onLayoutChanged);
+
+        emit canvasChanged();
+    }
+}
+
+float PDFSelection::wiggle() const
+{
+    return d->wiggle;
+}
+
+void PDFSelection::setWiggle(float newValue)
+{
+    if (newValue != d->wiggle) {
+        d->wiggle = newValue;
+        emit wiggleChanged();
+    }
+}
+
+void PDFSelection::Private::textBoxId(const QPointF &point, Position position,
+                                      int *pageIndex, int *boxIndex)
+{
+    *pageIndex = -1;
+    *boxIndex  = -1;
+    if (!canvas || !canvas->document())
+        return;
+    
+    // point is given in canvas coordinates.
+    QPair<int, QRectF> at = canvas->pageAtPoint(point);
+    if (at.first < 0)
+        return;
+    *pageIndex = at.first;
+
+    const PDFDocument::TextList &boxes = canvas->document()->textBoxesAtPage(*pageIndex);
+    switch (position) {
+    case PDFSelection::Private::At: {
+        *boxIndex = -1;
+        qreal squaredDistanceMin = wiggle * wiggle;
+        int i = 0;
+        for (PDFDocument::TextList::const_iterator box = boxes.begin();
+             box != boxes.end(); box++) {
+            qreal squaredDistance =
+                canvas->squaredDistanceFromRect(at.second, box->first, point);
+                
+            if (squaredDistance < squaredDistanceMin) {
+                *boxIndex = i;
+                squaredDistanceMin = squaredDistance;
+            }
+            i += 1;
+        }
+        break;
+    }
+    case PDFSelection::Private::Before: {
+        QPointF reducedCoordPoint {point.x() / at.second.width(),
+                (point.y() - at.second.y()) / at.second.height()};
+        *boxIndex = -1;
+        for (PDFDocument::TextList::const_iterator box = boxes.begin();
+             box != boxes.end(); box++) {
+            /* Stop counting boxes up as soon as a box after reducedCoordPoint is found. */
```

Comment:
Got a first confused with Before and After. Though they would mean like "give indexes for text box before this point", but suppose it's more like "give indexes for text boxes where point is before it".